jetty-util/src/main/java/org/eclipse/jetty/util/JavaVersion.java

```diff
@@ -34,115 +34,48 @@
     public static final String JAVA_TARGET_PLATFORM = "org.eclipse.jetty.javaTargetPlatform";
     
     /** Regex for Java version numbers */
-    private static final String VNUM = "(?<VNUM>[1-9][0-9]*(?:(?:\\.0)*\\.[0-9]+)*)";
-    private static final String UPDATE = "(?:(?<UNDERSCORE>_)(?<UPDATE>[0-9]+))?";
-    private static final String PRE = "(?:-(?<PRE>[a-zA-Z0-9]+))?";
-    private static final String BUILD = "(?:(?<PLUS>\\+)(?<BUILD>[0-9]+))?";
-    private static final String OPT = "(?:-(?<OPT>[-a-zA-Z0-9.~]+))?";
-
-    private static final String VSTR_FORMAT = VNUM + UPDATE + PRE + BUILD + OPT;
+    private static final String VSTR_FORMAT = "(?<VNUM>[1-9][0-9]*(?:(?:\\.0)*\\.[0-9]+)*).*";
 
     static final Pattern VSTR_PATTERN = Pattern.compile(VSTR_FORMAT);
     
-    public static final JavaVersion VERSION = parse(System.getProperty("java.runtime.version",System.getProperty("java.version")));
-
+    public static final JavaVersion VERSION = parse(System.getProperty("java.runtime.version",System.getProperty("java.version","1.8")));
```

Comment:
Invalid.

`java.version` is guaranteed to always be present in a JVM that follows the specification.

If `java.version` is empty/null then that should immediately be an error and Jetty should not start on this kind of JVM.

Even the Android dalvik and art VMs support `System.getProperty("java.version")`.

Also, the `java.runtime.version` is not right, its for the runtime's build identifier and is not tied to the JVM runtime details (which we care about).

We should use `java.version` only, no fallbacks.

Also, to make sure we are on spec compliant JVMs, here's the rules in the `java.version` parsing since Java 1.3.1 ...

> The content of the java.version system property must always be unique for each external binary.
> * A separator is always used between different parts of the string so that ordering can be determined programmically. (For example, a version string of "1.3.1beta" is not acceptable. A version string of "1.3.1-beta" is acceptable.)
> * The format should be numeric and in lower case where applicable with a period "." within the field itself and a dash "-" between fields.
> * A period "." separator indicates a major version.
> * An underbar "_" separator is used to indicate an update release.
> * A dash "-" separator is used to indicate a non-GA (non-FCS) release.
> * A GA (FCS) product version or build identifier must not contain a dash (-).
> * No asterik "*" or plus sign "+" is allowed, since these are used as wildcards for the matching logic.
> * All GA (FCS) versions are ordered based on the standard dot-notation. For example: 1.3.0 < 1.3.0_01 < 1.3.1 < 1.3.1_01.

Reply:
@joakime We know the rules, but we keep finding JVMs that break them.   Thus the intent of this change it to make jetty robust and never fail simply because of a versioning format error, or if some strange JVM doesn't provide "java.version". 

We are no longer parsing any of the detailed parts of "java.runtime.version", but we are using it if it is available at least for display purposes, so we log which JVM we are running on.   Hmmm having said that, I need to check if ALPN does need access to the update release version.

Reply:
ALPN uses the system property directly, so we are good there.

So I think this PR is good as it is less fragile that what we currently have.